Update cmake to 4.1.2 and start patching packages for compatibility

hnyman opened the pull request:
Reference to discussion in pr #19558 from @PolynomialDivision about updating tools/cmake to 4.x.

New cmake versions require at least 3.5 as 'cmake_minimum_required' in CMakeLists.txt of each compiled app.
Future cmake versions will increase that requirement to 3.10
https://cmake.org/cmake/help/latest/command/cmake_minimum_required.html

We have 64 cmake packages in the main OpenWrt repo and 200 packages in packages feed repo.
Some of those have already a new enough 'cmake_minimum_required', but many packages will need patching. As the CMakeLists.txt is part of upstream sources, we can't fully identify the affected packages from our local Makefiles.

The proper end-solution is naturally that the needed change is made to upstream sources, if possible. Many apps are already ok, and many are updating that 'cmake_minimum_required' value along their new versions.

I did this PR as an example of the needed changes and a wake-up call to OpenWrt core developers, who have direct access to the packages where "upstream" is actually an OpenWrt managed repo in OpenWrt git servers. They could rather easily patch upstream CMakeLists.txt while adding other new commits to upstream.  Active committers in those packages are e.g. @nbd168 @jow- @blogic @Ansuel 
(odhcpd was already changed a week ago, but I have not noticed other changes )

For "normal" packages with external upstream, we may need to patch them until upstream changes the value.
I propose that we locally patch it 3.30, which is the cmake version in the oldest currently supported OpenWrt branch (24.10). Main/master branch already uses cmake 3.31.

This likely requires somewhat similar action as the opkg-apk change, where quite many packages required versioning changes.

This PR has three commits:
* treewide: patch OpenWrt core apps for cmake 4.x compatibility  
  _Not comprehensive. Just the packages that were needed for my build_
- [x] mt76
- [x] libnl-tiny
- [x] libubox
- [x] uclient
- [x] udebug
- [x] ustream-ssl
- [x] odhcp6c
- [x] odhcpd
- [x] uhttpd
- [x] unetd
- [x] fstools
- [x] fwtool
- [x] procd
- [x] rpcd
- [x] ubox
- [x] ubus
- [x] ucert
- [x] uci
- [x] urngd
- [x] usign
- [x] ugps

* {jansson,libjson-c,jsonfilter}: patch for cmake 4.x compatibility
   _Not comprehensive. Just the packages that were needed for my build_

* tools/cmake: update to 4.1.2

Compiled / run-tested for mt7622/RT3200.

Ansuel commented:
@hnyman feedback about this: merge as soon as we have every package in core compilable. Then fix feeds later.

Ideally 25.xx will target this feature.

Ansuel commented:
For the core package I guess I can take care of updating the version and then bump the package... Do you want to create specific patch so I can merge with your sob and author tag?

hnyman commented:
I noticed that the advice in the cmake error message might actually offer a Makefile-only change possibility without a need for creating a patch for CMakeLists.txt.

Adding the value via CMAKE_OPTIONS in Makefile might be enough:
```
CMAKE_OPTIONS += \
	-DCMAKE_POLICY_VERSION_MINIMUM=3.10
```
That was enough to compile relayd, which is one of the core repo apps that I have not yet patched.
This might be more clean shortcut compared to separate patches for all OpenWrt core repo apps.
(Naturally the final solution is change at the "upstream" CMakeLists.txt

Does this sounds preferrable to (temporary) patches?

The error message is:

> CMake Error at CMakeLists.txt:1 (cmake_minimum_required):
>   Compatibility with CMake < 3.5 has been removed from CMake.
> 
>   Update the VERSION argument <min> value.  Or, use the <min>...<max> syntax
>   to tell CMake that the project requires at least <min> but has been updated
>   to work with policies introduced by <max> or earlier.
> 
>   Or, add -DCMAKE_POLICY_VERSION_MINIMUM=3.5 to try configuring anyway.

Ansuel commented:
Honestly I prefer to fix them directly than apply this workaround and
forget about it.

Il Ven 3 Ott 2025, 16:29 Hannu Nyman ***@***.***> ha scritto:

> *hnyman* left a comment (openwrt/openwrt#20265)
> <https://github.com/openwrt/openwrt/pull/20265#issuecomment-3365923591>
>
> I noticed that the advice in the cmake error message might actually offer
> a Makefile-only change possibility without a need for creating a patch for
> CMakeLists.txt.
>
> Adding the value via CMAKE_OPTIONS in Makefile might be enough:
>
> CMAKE_OPTIONS += \
> 	-DCMAKE_POLICY_VERSION_MINIMUM=3.10
>
> That was enough to compile relayd, which is one of the core repo apps that
> I have not yet patched.
> This might be more clean shortcut compared to separate patches for all
> OpenWrt core repo apps.
> (Naturally the final solution is change at the "upstream" CMakeLists.txt
>
> Does this sounds preferrable to (temporary) patches?
>
> The error message is:
>
> CMake Error at CMakeLists.txt:1 (cmake_minimum_required):
> Compatibility with CMake < 3.5 has been removed from CMake.
>
> Update the VERSION argument value. Or, use the ... syntax
> to tell CMake that the project requires at least but has been updated
> to work with policies introduced by or earlier.
>
> Or, add -DCMAKE_POLICY_VERSION_MINIMUM=3.5 to try configuring anyway.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/openwrt/openwrt/pull/20265#issuecomment-3365923591>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE2ZMQR3Z5IY3I77RHV5LT33V2B55AVCNFSM6AAAAACIA6OJPSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTGNRVHEZDGNJZGE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

hauke commented on package/network/services/odhcpd/files/odhcpd-update:
```diff
@@ -1,6 +1,8 @@
 #!/bin/sh
 # Make dnsmasq reread hostfile by sending SIGHUP signal
 
+logger -t odhcpd "*** ODHCPD triggers DNSMASQ reload ***"
```
How is this change related to cmake?

hnyman replied:
It isn't. Accidental inclusion in the commit. I will remove it.

Ansuel commented on package/libs/libjson-c/patches/002-cmake-version.patch:
```diff
@@ -0,0 +1,21 @@
++++ b/apps/CMakeLists.txt
+@@ -1,5 +1,5 @@
+ 
+-cmake_minimum_required(VERSION 2.8)  # see ../CMakeLists.txt for why 2.8
```
https://github.com/json-c/json-c/commit/d1716fe431694bfc08ad94d2e8d2a32d79b47fa3

Deprecated stuff so we can ignore.

Ansuel commented on tools/liblzo/patches/002-cmake-4.0-compatibility.patch:
```diff
@@ -0,0 +1,11 @@
+ #
+ 
+-cmake_minimum_required(VERSION 3.0 FATAL_ERROR)
++cmake_minimum_required(VERSION 3.30)
```
Oke thx

hnyman commented:
> Can you change to require 3.10?

I will do that on Sunday evening, and will also go through the main repo packages. There are still some unpatched packages, as they were not on my first list that you already patched. Nice that you got the main bulk away.

hnyman commented:
@Ansuel 
Regarding the already done changes, you missed libubox/lua sub-package.

hauke commented:
Please rebase, I fixed some other packages too.

Ansuel commented:
Tempted to merge and monitor faillog directory delta. @hauke what do you
think?

Il Dom 5 Ott 2025, 17:17 Hannu Nyman ***@***.***> ha scritto:

> *hnyman* left a comment (openwrt/openwrt#20265)
> <https://github.com/openwrt/openwrt/pull/20265#issuecomment-3369125901>
>
> I have now rebased it and also adjusted all changes to be into 3.10
> (instead of 3.30), as 3.10 was applied to the local core packages.
>
> I will still go through the main repo once more, as I think that there
> might still be a few external packages that still have too old version
> reference.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/openwrt/openwrt/pull/20265#issuecomment-3369125901>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE2ZMQQSDMEVKQ5FJXWYLY33WEZBNAVCNFSM6AAAAACIA6OJPSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTGNRZGEZDKOJQGE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

hnyman commented:
I checked that main repo, and did not find any more packages (or tools) to be fixed.
(Still evaluating llvm-bpf and dwarves, but all else should be 3.5+ after this PR)

hnyman commented:
Ps. I have already fixed the LuCI feed local packages.

hauke commented:
These packages are not building:
```
hauke@debian-16gb-nbg1-1:~/openwrt/openwrt$ grep "Compatibility with CMake < 3.5 has been removed from CMake" logs/ -r 
logs/package/feeds/packages/openobex/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/libsoxr/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/bluld/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/softethervpn5/host-compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/lualanes/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/realtek-poe/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/lua-cjson/lua-53/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/lua-cjson/lua-51/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/lua-cjson/lua-54/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/openelp/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/libyaml-cpp/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/base16384/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/leptonica/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/libftdi/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/uledd/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/mjpg-streamer/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/v4l2camera/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/libmad/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/telldus-core/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/zmq/curve/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/zmq/nc/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/libftdi1/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/mbusd/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/cjson/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/sysrepo/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/tini/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/canutils/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/tunneldigger/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/libnatpmp/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/quectel-timesync/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/usteer/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/rtpmidid/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/tiff/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/xxhash/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/hidapi/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/luv/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/liblo/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/libvorbis/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/libtins/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/lzmq/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/libsndfile/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/libyang/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/libid3tag/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/cereal/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/mbpoll/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/uci2/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/v4l2tools/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/nlbwmon/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/cgi-io/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/libuecc/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/libssh/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/portaudio/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/phantap/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/libroxml/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/qrencode/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/lua-ev/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/spawn-fcgi/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/zlog/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/ecdsautils/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/yajl/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/libwebsockets/mbedtls/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/libwebsockets/full/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/libwebsockets/openssl/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/respondd/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/msgpack-c/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/lzo/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/swanmon/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/protobuf/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/protobuf/host-compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/libradiotap/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/packages/v4l2rtspserver/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/routing/oonf-dlep-proxy/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/routing/oonf-dlep-radio/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/routing/oonf-olsrd2/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/routing/ohybridproxy/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/routing/minimalist-pcproxy/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/luci/lucihttp/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/telephony/sipp/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/telephony/bcg729/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
logs/package/feeds/video/libdouble-conversion/compile.txt:  Compatibility with CMake < 3.5 has been removed from CMake.
```

hnyman commented:
I noticed from CI that libevent2 had somehow slipped my check in main repo. 
Added now.

Regarding Hauke's test on feeds: Only 80 failures is rather ok. At the first glance there aren't that many really popular packages among failures.

Ansuel commented:
@hnyman maybe split this 0282d4acc9900e67db2b3187b2c35a24a4b09812 to 2 commit?

@hauke Did you notice other broken package in core repository? If it's not the case, I would merge this.

champtar commented:
lucihttp still need fixing https://github.com/jow-/lucihttp/issues/3 (luci-base -> liblucihttp-ucode)

hnyman committed:
tools/cmake: update to 4.1.2

Update cmake to 4.1.2.
Note that cmake 4.x requires at least 3.5 as 'cmake_minimum_required'
in CMakeLists.txt of each app to be compiled.

Future cmake versions will increase that requirement to 3.10
https://cmake.org/cmake/help/latest/command/cmake_minimum_required.html

Signed-off-by: Hannu Nyman <hannu.nyman@iki.fi>
Link: https://github.com/openwrt/openwrt/pull/20265
Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

hnyman committed:
libevent2: adjust for cmake 4.x compatibility

Patch CMakeLists.txt for cmake 4.x compatibility.
New cmake versions require at least 3.5 as 'cmake_minimum_required'
in CMakeLists.txt. In future 3.10 will be required.

Signed-off-by: Hannu Nyman <hannu.nyman@iki.fi>
Link: https://github.com/openwrt/openwrt/pull/20265
Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

hnyman committed:
tools/liblzo: add cmake 4.x compatibility

Patch CMakeLists.txt to fulfill cmake 4.0 requirements
of cmake_minimum_required being at least 3.5 and in future 3.10.

Signed-off-by: Hannu Nyman <hannu.nyman@iki.fi>
Link: https://github.com/openwrt/openwrt/pull/20265
Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

hnyman committed:
libjson-c: patch for cmake 4.x compatibility

Patch CMakeLists.txt in selected apps for cmake 4.x compatibility.
New cmake versions require at least 3.5 as 'cmake_minimum_required'
in CMakeLists.txt. In future 3.10 will be required.

Signed-off-by: Hannu Nyman <hannu.nyman@iki.fi>
Link: https://github.com/openwrt/openwrt/pull/20265
Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

hnyman committed:
jansson: adjust for cmake 4.x compatibility

Patch CMakeLists.txt in selected apps for cmake 4.x compatibility.
New cmake versions require at least 3.5 as 'cmake_minimum_required'
in CMakeLists.txt. In future 3.10 will be required.

Signed-off-by: Hannu Nyman <hannu.nyman@iki.fi>
Link: https://github.com/openwrt/openwrt/pull/20265
Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

hnyman committed:
treewide: edit local CMakeLists.txt to be compatible with cmake 4.x

Edit CMakeLists.txt to fulfill cmake 4.0 requirement that
cmake_minimum_required is now at least 3.5 and in future 3.10.

Signed-off-by: Hannu Nyman <hannu.nyman@iki.fi>
Link: https://github.com/openwrt/openwrt/pull/20265
Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

hnyman commented:
I split the one commit as suggested by @Ansuel and pushed the whole PR on top the current HEAD

kgoerbig commented:
These packages are not building in 6.12.57

[openwrt-imagebuilder-ramips-mt7621](https://downloads.openwrt.org/snapshots/targets/ramips/mt7621/openwrt-imagebuilder-ramips-mt7621.Linux-x86_64.tar.zst)

mjpg-streamer-input-uvc mjpg-streamer-output-http mjpg-streamer-www-simple kmod-video-uvc mjpg-streamer
